lib/floe/workflow/runner/docker.rb

```diff
@@ -18,34 +18,95 @@ def run!(resource, env = {}, secrets = {})
 
           image = resource.sub("docker://", "")
 
-          params  = ["run", :rm]
-          params += [[:net, "host"]] if network == "host"
-          params += env.map { |k, v| [:e, "#{k}=#{v}"] } if env
-
           secrets_file = nil
-
           if secrets && !secrets.empty?
-            secrets_file = Tempfile.new
-            secrets_file.write(secrets.to_json)
-            secrets_file.flush
+            secrets_file = create_secret(secrets)
+            env["_CREDENTIALS"] = "/run/secrets"
+          end
+
+          output = run_container(image, env, secrets_file)
+
+          [0, output]
+        ensure
+          secrets_file&.close!
+        end
+
+        def run_async!(resource, env = {}, secrets = {})
+          raise ArgumentError, "Invalid resource" unless resource&.start_with?("docker://")
 
-            params << [:e, "_CREDENTIALS=/run/secrets"]
-            params << [:v, "#{secrets_file.path}:/run/secrets:z"]
+          image = resource.sub("docker://", "")
+
+          secrets_file = nil
+          if secrets && !secrets.empty?
+            secrets_file = create_secret(secrets)
+            env["_CREDENTIALS"] = "/run/secrets"
           end
 
-          params << image
+          begin
+            container_id = run_container(image, env, secrets_file, :detached => true)
+          rescue
+            cleanup(container_id, secrets_file)
+            raise
+          end
 
-          logger.debug("Running docker: #{AwesomeSpawn.build_command_line("docker", params)}")
-          result = AwesomeSpawn.run!("docker", :params => params)
+          [container_id, secrets_file]
+        end
 
-          [result.exit_status, result.output]
-        ensure
+        def cleanup(container_id, secrets_file)
+          delete_container(container_id) if container_id
           secrets_file&.close!
         end
 
+        def running?(container_id)
+          inspect_container(container_id).first.dig("State", "Running")
+        end
+
+        def success?(container_id)
+          inspect_container(container_id).first.dig("State", "ExitCode") == 0
+        end
+
+        def output(container_id)
+          docker!("logs", container_id).output
+        end
+
         private
 
         attr_reader :network
+
+        def run_container(image, env, secrets_file, detached: false)
+          params  = ["run"]
+          params += detached ? [:detach] : [:rm]
+          params += [[:net, "host"]] if @network == "host"
+          params += env.map { |k, v| [:e, "#{k}=#{v}"] }
+          params << [:v, "#{secrets_file.path}:/run/secrets:z"] if secrets_file
+          params << image
```

Comment:
For consistency, thinking we should use `<<` everywhere.

```suggestion
          params  = ["run"]
          params << detached ? :detach : :rm
          params << [:net, "host"] if @network == "host"
          params += env.map { |k, v| [:e, "#{k}=#{v}"] }
          params << [:v, "#{secrets_file.path}:/run/secrets:z"] if secrets_file
          params << image
```

Reply:
I can `<<` the `--net host` one but
```
params << detached ? :detach : :rm
```

results in `false` not `:detach`
```
$ irb
>> detached = true
=> true
>> params = []
=> []
>> params << detached ? :detach : :rm
=> :detach
>> params
=> [true]
```

Reply:
Done

Reply:
Oh wild - I didn't realize that `<<` had priority over `?:`.

Only other thing would be the following, but it's up to you.

```ruby
params << (detached ? :detach : :rm)
```

Reply:
Sure that works

Reply:
Done